File: frontend/src/utils/translator.ts
import morseMap, { reverseMorseMap } from './morseMap'
import type {
  TranslationMode,
  TranslationResult,
  TranslationWarning,
  WarningCode,
} from '../types'

const morseTokenPattern = /^[.-]+$/
const maxWarningItems = 5

// choose the right singular or plural word for warning text.
function pluralize(count: number, singular: string, plural: string) {
  // warning messages reuse this so they read naturally: "1 token" vs "2 tokens".
  if (count === 1) {
    return singular
  }

  return plural
}

// normalize line endings once so decode/encode logic can treat newlines consistently.
function normalizeNewlines(value: string) {
  return value.replace(/\r\n?/g, '\n')
}

// look up the morse sequence for one plain-text character.
function getMorseCode(character: string) {
  const map = morseMap as Record<string, string>
  return map[character]
}

// look up the plain-text character for one full morse token.
function getDecodedCharacter(token: string) {
  const map = reverseMorseMap as Record<string, string>
  return map[token]
}

// show a few representative warning items without flooding the ui.
function getWarningItems(values: string[]) {
  const items: string[] = []

  for (const value of values) {
    if (items.includes(value)) {
      continue
    }

    items.push(value)

    if (items.length >= maxWarningItems) {
      break
    }
  }

  return items
}

// create one warning object in the format the ui expects.
function createWarning(
  code: WarningCode,
  message: string,
  items: string[],
): TranslationWarning {
  return {
    code,
    message,
    items: getWarningItems(items),
  }
}

// return the correct empty-input warning for the current mode.
function emptyWarning(mode: TranslationMode): TranslationWarning {
  if (mode === 'decode') {
    return {
      code: 'EMPTY_INPUT',
      message: 'Enter Morse code to decode.',
      items: [],
    }
  }

  return {
    code: 'EMPTY_INPUT',
    message: 'Enter text to encode.',
    items: [],
  }
}

// build the list of decode warnings after all tokens have been processed.
function buildDecodeWarnings(
  invalidSpacingTokens: string[],
  invalidTokens: string[],
  unknownTokens: string[],
) {
  const warnings: TranslationWarning[] = []

  if (invalidSpacingTokens.length > 0) {
    const count = invalidSpacingTokens.length
    const message =
      `Found invalid Morse spacing in ${count} ` +
      `${pluralize(count, 'section', 'sections')}; affected sections decoded as ?.`

    warnings.push(
      createWarning('INVALID_MORSE_SPACING', message, invalidSpacingTokens),
    )
  }

  if (invalidTokens.length > 0) {
    const count = invalidTokens.length
    const message =
      `Found invalid Morse characters in ${count} ` +
      `${pluralize(count, 'token', 'tokens')}; affected tokens decoded as ?.`

    warnings.push(
      createWarning('INVALID_MORSE_CHARACTERS', message, invalidTokens),
    )
  }

  if (unknownTokens.length > 0) {
    const count = unknownTokens.length
    const message =
      `Decoded ${count} unknown Morse ` +
      `${pluralize(count, 'token', 'tokens')} as ?.`

    warnings.push(
      createWarning('UNKNOWN_MORSE_TOKENS', message, unknownTokens),
    )
  }

  return warnings
}

// build the warning for unsupported text characters in encode mode.
function buildEncodeWarnings(unsupportedCharacters: string[]) {
  if (unsupportedCharacters.length === 0) {
    return []
  }

  const count = unsupportedCharacters.length
  const message =
    `Encoded ${count} unsupported ` +
    `${pluralize(count, 'character', 'characters')} as ?.`

  return [
    createWarning('UNSUPPORTED_TEXT_CHARACTERS', message, unsupportedCharacters),
  ]
}

// parse raw morse into word -> token groups.
// 1 space ends a letter, 3 spaces end a word, and any other spacing is preserved
// inside the token so decodeToken can flag it as invalid instead of silently guessing.
function splitDecodeWords(input: string) {
  const words: string[][] = []
  let currentWord: string[] = []
  let currentToken = ''
  let index = 0

  while (index < input.length) {
    const character = input[index]

    if (character !== ' ') {
      currentToken += character
      index += 1
      continue
    }

    let spaceCount = 0

    while (index < input.length && input[index] === ' ') {
      spaceCount += 1
      index += 1
    }

    if (spaceCount === 1) {
      if (currentToken !== '') {
        currentWord.push(currentToken)
        currentToken = ''
      }

      continue
    }

    if (spaceCount === 3) {
      if (currentToken !== '') {
        currentWord.push(currentToken)
        currentToken = ''
      }

      if (currentWord.length > 0) {
        words.push(currentWord)
        currentWord = []
      }

      continue
    }

    currentToken += ' '.repeat(spaceCount)
  }

  if (currentToken !== '') {
    currentWord.push(currentToken)
  }

  if (currentWord.length > 0) {
    words.push(currentWord)
  }

  return words
}

// decode one morse token and record why it failed if it cannot be decoded.
function decodeToken(
  token: string,
  invalidSpacingTokens: string[],
  invalidTokens: string[],
  unknownTokens: string[],
) {
  // tokens should arrive here as uninterrupted dot/dash sequences.
  // if spaces are still present, splitDecodeWords found malformed spacing.
  if (token.includes(' ')) {
    invalidSpacingTokens.push(token)
    return '?'
  }

  if (!morseTokenPattern.test(token)) {
    invalidTokens.push(token)
    return '?'
  }

  const decodedCharacter = getDecodedCharacter(token)

  if (!decodedCharacter) {
    unknownTokens.push(token)
    return '?'
  }

  return decodedCharacter
}

// decode a full morse string into plain text and collect any warnings along the way.
export function decodeMorse(input: string): TranslationResult {
  const normalizedInput = normalizeNewlines(input).trim()

  if (normalizedInput === '') {
    return {
      output: '',
      warnings: [emptyWarning('decode')],
    }
  }

  const invalidSpacingTokens: string[] = []
  const invalidTokens: string[] = []
  const unknownTokens: string[] = []
  // treat line breaks the same as morse word breaks so pasted multi-line input still works.
  const inputWithWordBreaks = normalizedInput.replace(/ *\n+ */g, '   ')
  const words = splitDecodeWords(inputWithWordBreaks)
  const decodedWords: string[] = []

  for (const word of words) {
    let decodedWord = ''

    for (const token of word) {
      const decodedCharacter = decodeToken(
        token,
        invalidSpacingTokens,
        invalidTokens,
        unknownTokens,
      )

      decodedWord += decodedCharacter
    }

    decodedWords.push(decodedWord)
  }

  return {
    output: decodedWords.join(' '),
    warnings: buildDecodeWarnings(
      invalidSpacingTokens,
      invalidTokens,
      unknownTokens,
    ),
  }
}

// encode plain text into morse, using 1 space between letters and 3 between words.
export function encodeText(input: string): TranslationResult {
  const normalizedInput = normalizeNewlines(input).trim()

  if (normalizedInput === '') {
    return {
      output: '',
      warnings: [emptyWarning('encode')],
    }
  }

  const unsupportedCharacters: string[] = []
  // collapse any whitespace run into a normal word boundary before encoding.
  const words = normalizedInput.split(/\s+/)
  const encodedWords: string[] = []

  for (const word of words) {
    if (word === '') {
      continue
    }

    const upperWord = word.toUpperCase()
    const encodedLetters: string[] = []

    for (const character of upperWord) {
      const encodedCharacter = getMorseCode(character)

      // keep the output aligned with the input instead of dropping unsupported characters.
      if (!encodedCharacter) {
        unsupportedCharacters.push(character)
        encodedLetters.push('?')
        continue
      }

      encodedLetters.push(encodedCharacter)
    }

    encodedWords.push(encodedLetters.join(' '))
  }

  return {
    output: encodedWords.join('   '),
    warnings: buildEncodeWarnings(unsupportedCharacters),
  }
}

File: backend/app/translator.py
import json
import re
from functools import lru_cache
from pathlib import Path

from .schemas import TranslationResponse, TranslationWarning

MORSE_TOKEN_PATTERN = re.compile(r"^[.-]+$")
MAX_WARNING_ITEMS = 5


def _pluralize(count: int, singular: str, plural: str) -> str:
    return singular if count == 1 else plural


def _collect_items(values: list[str]) -> list[str]:
    seen: dict[str, None] = {}
    for value in values:
        seen.setdefault(value, None)
    return list(seen.keys())[:MAX_WARNING_ITEMS]


def _empty_warning(mode: str) -> TranslationWarning:
    prompt = "Enter Morse code to decode." if mode == "decode" else "Enter text to encode."
    return TranslationWarning(code="EMPTY_INPUT", message=prompt)


def _normalize_newlines(value: str) -> str:
    return value.replace("\r\n", "\n").replace("\r", "\n")


def _split_decode_words(value: str) -> list[list[str]]:
    words: list[list[str]] = []
    current_word: list[str] = []
    current_token = ""
    index = 0

    def push_token() -> None:
        nonlocal current_token
        if not current_token:
            return

        current_word.append(current_token)
        current_token = ""

    def push_word() -> None:
        nonlocal current_word
        push_token()
        if not current_word:
            return

        words.append(current_word)
        current_word = []

    while index < len(value):
        if value[index] != " ":
            current_token += value[index]
            index += 1
            continue

        run_end = index
        while run_end < len(value) and value[run_end] == " ":
            run_end += 1

        run_length = run_end - index
        if run_length == 1:
            push_token()
        elif run_length == 3:
            push_word()
        else:
            current_token += " " * run_length

        index = run_end

    push_word()
    return words


@lru_cache(maxsize=1)
def load_morse_map() -> dict[str, str]:
    mapping_path = Path(__file__).resolve().parents[2] / "shared" / "morse-map.json"
    return json.loads(mapping_path.read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def load_reverse_morse_map() -> dict[str, str]:
    return {morse: character for character, morse in load_morse_map().items()}


def decode_morse(raw_input: str) -> TranslationResponse:
    normalized_input = _normalize_newlines(raw_input).strip()
    if not normalized_input:
        return TranslationResponse(
            mode="decode",
            input=raw_input,
            output="",
            warnings=[_empty_warning("decode")],
        )

    reverse_map = load_reverse_morse_map()
    invalid_spacing_tokens: list[str] = []
    invalid_tokens: list[str] = []
    unknown_tokens: list[str] = []
    expanded_input = re.sub(r" *\n+ *", "   ", normalized_input)
    words = _split_decode_words(expanded_input)
    decoded_words: list[str] = []

    for word in words:
        decoded_letters: list[str] = []
        for token in word:
            if " " in token:
                invalid_spacing_tokens.append(token)
                decoded_letters.append("?")
                continue

            if not MORSE_TOKEN_PATTERN.fullmatch(token):
                invalid_tokens.append(token)
                decoded_letters.append("?")
                continue

            decoded_letter = reverse_map.get(token)
            if decoded_letter is None:
                unknown_tokens.append(token)
                decoded_letters.append("?")
                continue

            decoded_letters.append(decoded_letter)

        decoded_words.append("".join(decoded_letters))

    warnings: list[TranslationWarning] = []

    if invalid_spacing_tokens:
        count = len(invalid_spacing_tokens)
        warnings.append(
            TranslationWarning(
                code="INVALID_MORSE_SPACING",
                message=(
                    f"Found invalid Morse spacing in {count} "
                    f"{_pluralize(count, 'section', 'sections')}; affected sections decoded as ?."
                ),
                items=_collect_items(invalid_spacing_tokens),
            )
        )

    if invalid_tokens:
        count = len(invalid_tokens)
        warnings.append(
            TranslationWarning(
                code="INVALID_MORSE_CHARACTERS",
                message=(
                    f"Found invalid Morse characters in {count} "
                    f"{_pluralize(count, 'token', 'tokens')}; affected tokens decoded as ?."
                ),
                items=_collect_items(invalid_tokens),
            )
        )

    if unknown_tokens:
        count = len(unknown_tokens)
        warnings.append(
            TranslationWarning(
                code="UNKNOWN_MORSE_TOKENS",
                message=(
                    f"Decoded {count} unknown Morse "
                    f"{_pluralize(count, 'token', 'tokens')} as ?."
                ),
                items=_collect_items(unknown_tokens),
            )
        )

    return TranslationResponse(
        mode="decode",
        input=raw_input,
        output=" ".join(decoded_words),
        warnings=warnings,
    )


def encode_text(raw_input: str) -> TranslationResponse:
    normalized_input = _normalize_newlines(raw_input).strip()
    if not normalized_input:
        return TranslationResponse(
            mode="encode",
            input=raw_input,
            output="",
            warnings=[_empty_warning("encode")],
        )

    morse_map = load_morse_map()
    unsupported_characters: list[str] = []
    encoded_words: list[str] = []

    for word in [part for part in re.split(r"\s+", normalized_input) if part]:
        encoded_letters: list[str] = []
        for character in word.upper():
            encoded_token = morse_map.get(character)
            if encoded_token is None:
                unsupported_characters.append(character)
                encoded_letters.append("?")
                continue

            encoded_letters.append(encoded_token)

        encoded_words.append(" ".join(encoded_letters))

    warnings: list[TranslationWarning] = []

    if unsupported_characters:
        count = len(unsupported_characters)
        warnings.append(
            TranslationWarning(
                code="UNSUPPORTED_TEXT_CHARACTERS",
                message=(
                    f"Encoded {count} unsupported "
                    f"{_pluralize(count, 'character', 'characters')} as ?."
                ),
                items=_collect_items(unsupported_characters),
            )
        )

    return TranslationResponse(
        mode="encode",
        input=raw_input,
        output="   ".join(encoded_words),
        warnings=warnings,
    )
